Reader datasets from R language

weightwatchers-carlanderson opened the pull request:
This is a simple reader that reads in datasets from R.
It is setup as an optional install
The test suite is set up as optional too, looking for an rpy2 install

@kzecchini @parul-l I didn't want to replicate the documentation in https://github.com/ww-tech/primrose/pull/27/files 
so whichever of these 2 PRs is merged first, we can add a couple more lines of documentation top say that R is also an optional install in the 2nd PR.

parul-l commented on test/test_r_reader.py:
```diff
@@ -0,0 +1,59 @@
+from primrose.base.node import AbstractNode
+from primrose.data_object import DataObject, DataObjectResponseType
+
+@pytest.mark.skipif('rpy2' in sys.modules, reason="primrose[R] is optional")
```
typo: `.skipif(rpy2 is not in sys.modules...)`
This seems like a simpler way of doing things. In my PR, I created a [custom marker](http://doc.pytest.org/en/latest/example/markers.html) and then called the relevant pytest command in the `.travis.yml` file.  Not sure which is the best method for testing purposes, but I think this way is better. I'll see if others have comments on my PR as well.

weightwatchers-carlanderson replied:
thanks for catching this @parul-l. This is now fixed.

weightwatchers-carlanderson replied:
@kzecchini do you have an opinion on this? Which is the better pattern to define the test skips?

kzecchini replied:
sorry I missed this thread earlier! I think that the `skipif` here is preferable just so it is easier to test in travis. We can just run a single command of `pytest` and internally it will figure out what to do with the tests. If we use custom markers then we need to have multiple different pytest calls with different markers, for example `pytest -m "not plotting"` `pytest -m "not postgres"`. This might be harder to manage.

For now we I am leaving the other marks as is, but it could be an improvement to standardize to the `skipif` method.

parul-l replied:
@kzecchini @weightwatchers-carlanderson sounds good. I'll create an issue to adjust the test structure to use `skipif`. I think you're right - it is clearer this way.

parul-l commented on test/test_r_reader.py:
```diff
@@ -0,0 +1,59 @@
+    assert df.shape == (150, 6)
+    assert list(df.columns) == ["Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width", "Species", "row_names"]
+
+@pytest.mark.skipif('rpy2' in sys.modules, reason="primrose[R] is optional")
```
typo: `.skipif(rpy2 is not in sys.modules...)`

weightwatchers-carlanderson replied:
This is now fixed

parul-l commented:
Hey @weightwatchers-carlanderson. Is the build failing because you need to pull the latest version of the `.travis.yaml` file? I think other than that, everything looks good.

weightwatchers-carlanderson commented:
> Hey @weightwatchers-carlanderson. Is the build failing because you need to pull the latest version of the `.travis.yaml` file? I think other than that, everything looks good.

Hi @parul-l if your PR 27 was already merged, then yes this PR will be out of date (I have not touched this for weeks) and needs to updated from master. I will also want to update the docs to add this new feature to your docs. I'll look into it.

weightwatchers-carlanderson commented:
@parul-l @kzecchini I rebased from master, I fixed a typo in my test/test_r_reader unit test which now passes, and I added some docs about primrose[R]. All checks have now passed.

kzecchini commented:
I think overall this looks good. We just need to update the `.travis.yml` file to include the new `R` optional install. Right now it looks like we don't install `rpy2` or test this reader.

Not sure if it should be in this PR, but I also think we should have a `primrose[all]` command which installs all of our dependencies if someone wants the entire package.

I can make this change quickly and add to the PR.

parul-l commented:
> I think overall this looks good. We just need to update the `.travis.yml` file to include the new `R` optional install. Right now it looks like we don't install `rpy2` or test this reader.
hey @kzecchini ...i'm just confused here. isn't the `test/test_r_reader.py` file testing the reader? are you referring to something else?

kzecchini commented:
@parul-l It would skip every time - it looks like you only have `rpy2` in your `sys.modules` if it is imported. Since it is never imported it is never tested. I also realized you need `R` installed on travis, so I am figuring that out now. Tests are failing currently when actually running the r reader tests.

parul-l commented:
> @parul-l It would skip every time - it looks like `rpy2` is only in your `sys.modules` if it is imported. Since it is never imported it is never tested. I also realized you need `R` installed on travis, so I am figuring that out now. Tests are failing currently when actually running the r reader tests.

I see. thanks for the clarification.

kzecchini commented:
@parul-l I figured out how to get r installed properly. I think it would be best if you gave the final approval for this one since I made some commits! Thanks!

kzecchini commented:
@parul-l Followed the instructions for install on linux here: https://cran.cnr.berkeley.edu/

Right now it is hardcoded lines in our build to match with ubuntu xenial for travis build and match with debian stretch for the docker build. They have the keys listed on the site and are needed to download from the source via `apt-get`
